### PR TITLE
Add `@define` annotation to propagate linktime reachability to native code compile phase

### PR DESCRIPTION
**Pull request by armanbilge**
Inspired by @ekrich in https://github.com/scala-native/scala-native/pull/3389. Fixes https://github.com/scala-native/scala-native/issues/2778.

This PR introduces a new `@define` annotation. It is similar to the `@link` annotation which automatically adds an `-l` flag to the linker options if an extern object is linktime-reachable. In this case, `@define` adds a `-D` macro definition flag to the compiler options if the object is linktime-reachable.

**In short: this annotation enables us to propagate Scala linktime reachability to the native code compile phase.**

Furthermore, it does so independently of the `@link` annotation: it is an orthogonal feature that may be useful in other situations.

In nativelib we now use `@define("SCALANATIVE_Z")` to make compilation of zlib glue code conditional on its linktime reachability.

All naming subject to bikeshed: "define" is a bit overloaded in the codebase so in some places I refer to these as "macros" as in "macro definitions". So we might want to make this more consistent.

### CODE REVIEW

**Comment by armanbilge** [comment]
I suppose we could generalize this even further. Instead of having `@link` and `@define` we could have e.g. `@nativeOptions(compileFlags, linkFlags)`: "if this code is reachable at linktime, add these compile flags and add these linking flags". 🤔

**Comment by ekrich** [comment]
There are a couple of issues I see with this proposal.

1. As far as I know the `linkerResult`is global so every `NativeLib` will get compiled with the flags even if they are not needing the defines.
2. Not having a set pattern for the defines means that different libraries could conflict with each other.
3.  I can't see how this is simpler than the other proposal.

The other proposal is designed to be project by project and it gives us the potential in the future to add other flags that can be used for compilation that are only for the dependency being compiled. It also exactly fixes the current issue at hand.

**Comment by armanbilge** [comment]
1. Why is this harmful?

2. There is also no set pattern for naming of glue methods, which may also conflict. We already rely on developers to do the right thing here. This is nothing new.

3. It's just my opinion, and perhaps I misunderstood your proposal. But from my perspective involves many moving parts:

    - a properties file with **additional configuration that is redundant** e.g. you need both `@link("z")` and `nir.link.names = z`.

    - references to the group and artifact ids within said properties file. **I fear we are hard-coding too much about the maven model in the Scala Native toolchain**. It should just be working with ordinary jars/classpath without having to know about group ids and such.
 
   - tanglement of the macro definitions with the `@link` annotation. As demonstrated here, these features are actually orthogonal. `extern` methods and glue code don't have to be connected with a library to `@link` to. There may be other situations where conditional compilation based on linktime reachability is useful, for example when calling system APIs.

**Comment by ekrich** [comment]
The maven coordinates are well known and unique for dependencies. They are only used to disambiguate the library that contains the GC so the special logic can be used for GC compilation. It could be used for compile logging but there are no planned usages at this time. It is not mentioned in the docs so that is an internal detail

The duplication of the link `z` for example is opt-in if you want to use the feature - it only applies to the dependency being compiled. Conceivable, more that one library could use `libz`. This purposely keeps the front end NIR compilation/optimization decoupled from the backend C compilation as an opt-in feature. NIR to linking makes sense because that is global for the application.

Adding the ability for the library to specify options. Currently we compile posixlib with no conformance because we need defines. Currently we compile javalib with `gnu11` which automatically sets all sorts of gnuisms and C lib extension. That is plainly wrong.

Talking about a build system is not correct. The build system is the tools process which compiles files and then links them. We are not talking about build dependencies and other things for a full build system, we are just talking about having the ability for libraries to have some control over how they are built.

I am purposely taking thing very slowly with the future in mind and I think the descriptor approach is better because it potentially has longer legs. It solves the problem at hand as well.

**Comment by armanbilge** [comment]
Thanks for the follow up. I think we might just have to agree to disagree :)

---

> The maven coordinates are well known and unique for dependencies.

Right, I'm not objecting to that :) indeed, _if_ we use Maven coordinates, they would be well-known and unique, which is good.

I'm objecting to using Maven coordinates _at all._ I don't think the Maven coordinate scheme should be hard-coded into the Scala Native tools. The toolchain should be based on classpaths without any knowledge of an artifactory scheme.

---

> Talking about a build system is not correct ... we are just talking about having the ability for libraries to have some control over how they are built.

This sounds a lot like a build system to me 😅

**Comment by ekrich** [comment]
The maven stuff is not in the documentation and is only used to determine which library has the GC rather than just the file as before.

It is not a "build system" even if it allows a library developer to supply compile options. That is because we control which options and parts of the compile command that they could control. I gave some good examples for POSIX conformance and compiler like C11 vs GNU11 which are real concerns. Besides none of that is on the table at this time but even if it was, it is not a "build system" outside of what Scala Native tools allows.

The other proposal still would allow the developer to supply link targets as suggested by Wojciech https://github.com/scala-native/scala-native/pull/3389#issuecomment-1665419269 The defaults link could still be left in the code but the link targets not added to the link command when the option is chosen.

The lack of having control over naming where it could affect other dependencies and the defines being global to all native libs is still a concern.

**Comment by tanishiking** [comment]
Personally, I agree with the `@define' annotation (although there may be a better name for it).

Ekrich's descriptor idea https://github.com/scala-native/scala-native/issues/2778 is very interesting, and I found the idea of controlling the conditional compilation of glue code depending on the reachability of certain functions provided by the Scala library (which uses a C library) very useful.

However, I also agree with Arman that Scala's native toolchains should not be overly involved with configuring flags, and the only original PR concern for me was where this annotation would add a linker flag.

So this change (which, IIUC, is a version that does not affect the linker flag of ekrick's change) seems like just the right solution to me.
The linker flag would then be provided by the user.

---

[Below is just saying, can skip it 😃 ]
One thing that comes to mind is Cargo's feature flag: https://doc.rust-lang.org/cargo/reference/features.html#the-features-section 

It would be nice if scala-native or something similar could provide similar functionality, but that would probably mean scala-native moving closer to its own build system.

A native approach would be for the library author to do conditional compilation in the glue code, and include a clause in the README saying that if you want to opt in or out of some features (using the C library), you should use these compilation options (e.g. `LINK_LIB_Z` or something like that).
I thought it would be enough to write a manual, but it might be too optimistic to think that such a very native approach would work in the community.

**Review by WojciechMazur — COMMENTED** [review]
Looks nice, I've got just one naming comment and one security issue to discuss. 

I really like how easy it makes to adjust the compilation process. I enjoy the fact that all that is required could be defined in the source code. However, I'm a bit unsecure about about leaking possibly harmfull defintions to other compilation units.  I have an idea how to fix that with a workaround allowing to create a sandbox for compilation environment for each compilation unit (dependency with native sources). 
This way we would be able to combine the simpliciy of this solution with limiting the scope of affected sources similary to the proposal of Eric in his alternative solution.

File: nscplugin/src/main/scala-2/scala/scalanative/nscplugin/NirGenStat.scala [hunk 1]
```diff
@@ -157,6 +157,19 @@ trait NirGenStat[G <: nsc.Global with Singleton] { self: NirGenPhase[G] =>
...
           Attr.Link(name)
+        case ann if ann.symbol == DefineClass =>
+          (ann.tree: @unchecked) match {
+            case Apply(_, Seq(Literal(Constant(name: String)))) =>
+              Attr.Define(name, None)
+            case Apply(
+                  _,
+                  Seq(
+                    Literal(Constant(name: String)),
+                    Literal(Constant(definition: String))
```
[the remark this answers is not in the document]
Comment (ekrich) [reply nscplugin/src/main/scala-2/scala/scalanative/nscplugin/NirGenStat.scala:168]: But doesn't this all happen at the global NIR level and these are going to to happen for every compiled C file? My PR limits the settings to only the current dependency being compiled that has native code. The GC define only happens for the nativelib so they apply library by library - in the long run we need that for all the reasons I have explained in comments of one or the other PRs.

[the remark this answers is not in the document]
Comment (ekrich) [reply nscplugin/src/main/scala-2/scala/scalanative/nscplugin/NirGenStat.scala:168]: The other thing that a delayed Descriptor can do is allow the build to inspect options and do different things based on the options which could be helpful for the example of linking to different frameworks. To encode this stuff in annotations is probably worse than in a descriptor and then you will also have to have all the logic in the build either way as that is where the logic occurs.

[the remark this answers is not in the document]
Comment (armanbilge) [reply nscplugin/src/main/scala-2/scala/scalanative/nscplugin/NirGenStat.scala:168]: Ahh, thanks, I didn't think about this. I think for now it should be sufficient to have simple macro definitions without custom values. We can revisit that feature in the future if we find a way to scope the compiler options to the relevant glue codes, like Eric's PR does. I do think that is a good idea but it would be nice if it could happen "automatically" somehow.

**Comment by ekrich** [comment]
Come to think about it, we could just add a define for all the linked libs that show up via NIR
`s"-DSCALANATIVE_LINK_${name.toUpperCase}"` and not have the `@define` at all and then the potential for conflict is less. The user just adds the conditional compilation in code. Either way to avoid the error when not used the library author has to match the define.

If you can somehow scope `@link` to a library/project and then somehow figure out in compile backend which library you are compiling and only add the ones for the project that would work but then it is not optional.

The other proposal is targeted to the project, relatively safe (no user defines), and optional. I am not seeing the upside other than not reading a properties file that says that you want a conditional compilation for the libraries you specify which is a duplication of the `@link`.

**Comment by armanbilge** [comment]
Yes we could, but I think it's more advantageous for `@define` to be an independent feature. For example here are some interesting things you can do:

```scala
@link("curl")
@define("HTTP4S_CURL")
@extern
object curl {
  // curl functions
}

@link("curl")
@define("HTTP4S_CURL_WS")
@extern
object curl_ws {
 // websocket specific functions
}
```

cURL has [experimental websocket support](https://curl.se/libcurl/c/libcurl-ws.html). To use websocket functions the user must compile cURL with a special flag to enable experimental websocket support.

By guarding websocket feature behind a separate definition `HTTP4S_CURL_WS` you can use the library with a vanilla build of cURL (no websockets available) as long as websocket related functions are not linktime-reachable.

In summary, this means that a Scala library can gracefully support extra features when linking to newer/experimental versions of the same native C library, depending on the linktime reachability of the extern functions that the user calls.

**Comment by ekrich** [comment]
In your example above about supporting different curl versions, they recommend the following which could be used for conditional compilation. This way you shouldn't need 2 different defines.

https://curl.se/docs/versions.html

**Comment by armanbilge** [comment]
> In your example above about supporting different curl versions

It's not about different curl versions. It's about if curl is compiled with experimental websockets enabled or not. This is not related to the version of curl, but rather the configuration that was used when it was compiled.

Regardless, my point is that an independent `@define` annotation gives us the _flexibility_ to do things like that as needed.

**Comment by WojciechMazur** [comment]
Thank you for this PR! I'd like to merge it to the main branch. If that's not a problem, I'm going to rebase it on current HEAD and resolve conflicts. I hope you don't mind if I would need to force push here
